### PR TITLE
Enable Web process cache on iOS

### PR DESCRIPTION
**Pull request by pvollan**
#### 7d015b7c0f7a00805f719bf96e04f2f33b2ff77d
<pre>
Enable Web process cache on iOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=258978">https://bugs.webkit.org/show_bug.cgi?id=258978</a>
rdar://111908431

Reviewed by Brent Fulgham.

Enable the Web process cache on iOS by addressing an issue where the physical memory size in GB is reported as zero,
causing the process cache capacity to be set to zero. This is because the jetsam limit is taken into consideration
in this computation on iOS. This patch adds a function that returns the memory size without considering the jetsam
limit of the calling process. This patch also delays releasing memory when preparing to suspend, since this is
further improving the page load speed when reusing a process from the cache. Releasing the memory is delayed until
we drop the last assertion, which will suspend the process. Until then, the process is running and able to respond
to memory pressure warnings from the system and releasing memory. Furthermore, the process is not added to the cache
if assertions are being held, preventing it from suspending. This is a fix for a regression in page load speed we
see on <a href="http://www.youtube.com">http://www.youtube.com</a> when adding it to the cache. This change is a speedup on page load benchmarks.

* Source/WTF/wtf/RAMSize.cpp:
(WTF::ramSizeDisregardingJetsamLimit):
* Source/WTF/wtf/RAMSize.h:
* Source/WebKit/UIProcess/WebProcessCache.cpp:
(WebKit::WebProcessCache::updateCapacity):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
* Source/WebKit/WebProcess/WebProcess.cpp:

Canonical link: <a href="https://commits.webkit.org/266267@main">https://commits.webkit.org/266267@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/46fa1a4d4f97e4909d4be44b3144495699460fe9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13374 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13688 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14020 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15111 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12732 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16195 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13714 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15414 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13541 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14195 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11314 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15764 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11483 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12069 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/19119 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/11398 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12558 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/12236 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/15450 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/12662 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12739 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/10599 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/13419 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12011 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3515 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3265 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16333 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/13803 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12580 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3312 "Passed tests") |

### CODE REVIEW

File: Source/WebKit/UIProcess/WebProcessCache.cpp [hunk 1]
```diff
@@ -41,6 +41,30 @@ Seconds WebProcessCache::cachedProcessLifetime { 30_min };
 Seconds WebProcessCache::clearingDelayAfterApplicationResignsActive { 5_min };
 static Seconds cachedProcessSuspensionDelay { 30_s };
 
+#if PLATFORM(IOS)
+static size_t ramSizeWithoutJetsamLimit()
+{
+    host_basic_info_data_t hostInfo;
+
+    mach_port_t host = mach_host_self();
```
Comment (brentfulgham) [inline Source/WebKit/UIProcess/WebProcessCache.cpp]: It's a shame we don't have a smart pointer for this type.

File: Source/WebKit/UIProcess/WebProcessCache.cpp [hunk 2]
```diff
@@ -41,6 +41,30 @@ Seconds WebProcessCache::cachedProcessLifetime { 30_min };
...
+    mach_port_deallocate(mach_task_self(), host);
+    if (r != KERN_SUCCESS)
+        return 0;
+
+    size_t memSize = static_cast<size_t>(hostInfo.max_mem);
+    
+    if (hostInfo.max_mem > std::numeric_limits<size_t>::max())
+        memSize = std::numeric_limits<size_t>::max();
+
+    size_t multiple = 128 * MB;
```
Comment (brentfulgham) [inline Source/WebKit/UIProcess/WebProcessCache.cpp]: It seems like this could be a `constexpr` outside of the method.

File: Source/WebKit/UIProcess/WebProcessCache.cpp [hunk 3]
```diff
@@ -41,6 +41,30 @@ Seconds WebProcessCache::cachedProcessLifetime { 30_min };
...
+    mach_msg_type_number_t count = HOST_BASIC_INFO_COUNT;
+    kern_return_t r = host_info(host, HOST_BASIC_INFO, (host_info_t)&hostInfo, &count);
+    mach_port_deallocate(mach_task_self(), host);
+    if (r != KERN_SUCCESS)
+        return 0;
+
+    size_t memSize = static_cast<size_t>(hostInfo.max_mem);
+    
+    if (hostInfo.max_mem > std::numeric_limits<size_t>::max())
+        memSize = std::numeric_limits<size_t>::max();
```
Comment (brentfulgham) [inline Source/WebKit/UIProcess/WebProcessCache.cpp]: `memSize = std::min(hostInfo.max_mem, std::numeric_limits<size_t>::max())`

File: Source/WebKit/UIProcess/WebProcessCache.cpp [hunk 4]
```diff
@@ -41,6 +41,30 @@ Seconds WebProcessCache::cachedProcessLifetime { 30_min };
...
+
+    size_t multiple = 128 * MB;
+
+    // Round up the memory size to a multiple of 128MB
+    return ((memSize + multiple - 1) / multiple) * multiple;
```
Comment (brentfulgham) [inline Source/WebKit/UIProcess/WebProcessCache.cpp]: `return divideRoundedUp(memorize, multiple) * multiple;`

Comment (brentfulgham) [reply Source/WebKit/UIProcess/WebProcessCache.cpp]: Might need to `#include <wtf/MathExtras.h>`

File: Source/WebKit/UIProcess/WebProcessCache.cpp [hunk 5]
```diff
@@ -128,7 +152,7 @@ bool WebProcessCache::addProcess(std::unique_ptr<CachedProcess>&& cachedProcess)
     RELEASE_ASSERT(!registrableDomain.isEmpty());
 
     if (auto previousProcess = m_processesPerRegistrableDomain.take(registrableDomain))
-        WEBPROCESSCACHE_RELEASE_LOG("addProcess: Evicting process from WebProcess cache because a new process was added for the same domain", previousProcess->process().processID());
+        WEBPROCESSCACHE_RELEASE_LOG("addProcess: Evicting process from WebProcess cache because a new process was added for the same domain %{public}s", previousProcess->process().processID(), registrableDomain.string().utf8().data());
```
Comment (brentfulgham) [inline Source/WebKit/UIProcess/WebProcessCache.cpp]: I think we treat `registrableDomain` as `{private}`, because we don't want logging of origins visited.

File: Source/WebKit/UIProcess/WebProcessCache.cpp (hunk 1 above)
Comment (brentfulgham) [reply Source/WebKit/UIProcess/WebProcessCache.cpp]: Maybe `ramSizeWithoutJetsamLimit` should just be added to the `wtf/RamSize.{h,cpp}` file?

File: Source/WebKit/UIProcess/WebProcessCache.cpp (hunk 5 above)
Comment (pvollan) [reply Source/WebKit/UIProcess/WebProcessCache.cpp]: Thanks for reviewing! I will address these issues in the coming patch.

File: Source/WTF/wtf/RAMSize.cpp [hunk 6]
```diff
@@ -82,4 +87,43 @@ size_t ramSize()
...
+#if PLATFORM(IOS)
+size_t ramSizeDisregardingJetsamLimit()
+{
+    host_basic_info_data_t hostInfo;
+
+    mach_port_t host = mach_host_self();
+    mach_msg_type_number_t count = HOST_BASIC_INFO_COUNT;
+    kern_return_t r = host_info(host, HOST_BASIC_INFO, (host_info_t)&hostInfo, &count);
+    mach_port_deallocate(mach_task_self(), host);
+    if (r != KERN_SUCCESS)
```
Comment (brentfulgham) [inline Source/WTF/wtf/RAMSize.cpp:99]: Perhaps `deallocateSendRightSafely()` would be better? Do we really want to just keep going if the mach port is in a bad state?

File: Source/WebKit/UIProcess/WebProcessCache.cpp [hunk 7]
```diff
@@ -59,8 +63,38 @@ WebProcessCache::WebProcessCache(WebProcessPool& processPool)
     platformInitialize();
 }
 
-bool WebProcessCache::canCacheProcess(WebProcessProxy& process) const
+bool WebProcessCache::canCacheProcess(WebProcessProxy& process, bool checkSuspendedState) const
```
Comment (brentfulgham) [inline Source/WebKit/UIProcess/WebProcessCache.cpp]: The boolean would be better as a two-value Enum, `CheckSuspendedState::Yes`

File: Source/WebKit/UIProcess/WebProcessCache.cpp [hunk 8]
```diff
@@ -91,7 +125,7 @@ bool WebProcessCache::addProcessIfPossible(Ref<WebProcessProxy>&& process)
     ASSERT(!process->suspendedPageCount());
     ASSERT(!process->isRunningServiceWorkers());
 
-    if (!canCacheProcess(process))
+    if (!canCacheProcess(process, true))
```
Comment (brentfulgham) [inline Source/WebKit/UIProcess/WebProcessCache.cpp]: `if (!canCacheProcess(process, CheckSuspendedState::Yes))`

File: Source/WebKit/UIProcess/WebProcessCache.cpp [hunk 9]
```diff
@@ -121,7 +154,7 @@ bool WebProcessCache::addProcess(std::unique_ptr<CachedProcess>&& cachedProcess)
     ASSERT(!cachedProcess->process().suspendedPageCount());
     ASSERT(!cachedProcess->process().isRunningServiceWorkers());
 
-    if (!canCacheProcess(cachedProcess->process()))
+    if (!canCacheProcess(cachedProcess->process(), false))
```
Comment (brentfulgham) [inline Source/WebKit/UIProcess/WebProcessCache.cpp]: `if (!canCacheProcess(cachedProcess->process(), CheckSuspendedState::No)`

File: Source/WebKit/UIProcess/WebProcessCache.cpp [hunk 10]
```diff
@@ -174,24 +207,42 @@ RefPtr<WebProcessProxy> WebProcessCache::takeProcess(const WebCore::RegistrableD
 
 void WebProcessCache::updateCapacity(WebProcessPool& processPool)
 {
-    if (!processPool.configuration().processSwapsOnNavigation() || !processPool.configuration().usesWebProcessCache() || LegacyGlobalSettings::singleton().cacheModel() != CacheModel::PrimaryWebBrowser || processPool.configuration().usesSingleWebProcess()) {
+#if PLATFORM(IOS)
+    static bool hasEnabledProcessCache = false;
+    bool usesWebProcessCache = isFullWebBrowserOrRunningTest() && !hasEnabledProcessCache;
```
Comment (brentfulgham) [inline Source/WebKit/UIProcess/WebProcessCache.cpp]: Is `WebProcessCache::updateCapacity` called very frequently? I think `isFullWebBrowserOrRunningTest()` is somewhat costly, and doesn't change during the life of the UIProcess. Could this be a run-once pattern instead?

Comment (brentfulgham) [reply Source/WebKit/UIProcess/WebProcessCache.cpp]: Also: Is this behavior needed in Catalyst?

File: Source/WTF/wtf/RAMSize.cpp (hunk 6 above)
Comment (pvollan) [reply Source/WTF/wtf/RAMSize.cpp:99]: That is a good point. I added an error check and early return.

Thanks for reviewing!

File: Source/WebKit/UIProcess/WebProcessCache.cpp (hunk 7 above)
Comment (pvollan) [reply Source/WebKit/UIProcess/WebProcessCache.cpp]: Fixed!

File: Source/WebKit/UIProcess/WebProcessCache.cpp (hunk 8 above)
Comment (pvollan) [reply Source/WebKit/UIProcess/WebProcessCache.cpp]: Done.

File: Source/WebKit/UIProcess/WebProcessCache.cpp (hunk 9 above)
Comment (pvollan) [reply Source/WebKit/UIProcess/WebProcessCache.cpp]: Done.

File: Source/WebKit/WebProcess/WebProcess.cpp [hunk 11]
```diff
@@ -447,7 +448,7 @@ void WebProcess::initializeWebProcess(WebProcessCreationParameters&& parameters)
             if (m_pagesInWindows.isEmpty() && critical == Critical::No)
                 critical = Critical::Yes;
 
-#if PLATFORM(MAC)
+#if PLATFORM(MAC) || PLATFORM(IOS)
```
Comment (brentfulgham) [inline Source/WebKit/WebProcess/WebProcess.cpp]: Should this be `#if PLATFORM(COCOA)`? Or do you really mean the Catalyst, watchOS, etc., should not get this logic?

File: Source/WTF/wtf/RAMSize.cpp [hunk 12]
```diff
@@ -82,4 +87,44 @@ size_t ramSize()
...
+
+    return static_cast<size_t>(hostInfo.max_mem);
+}
+
+size_t ramUnused()
+{
+    mach_msg_type_number_t hostSize;
+    vm_size_t pageSize;
+
+    hostSize = sizeof(vm_statistics_data_t) / sizeof(integer_t);
```
Comment (brentfulgham) [inline Source/WTF/wtf/RAMSize.cpp]: I much prefer `mach_msg_type_number_t hostSize = sizeof(vm_statistics_data_t) / sizeof(integer_t);`

File: Source/WebKit/UIProcess/WebProcessCache.cpp (hunk 10 above)
Comment (pvollan) [reply Source/WebKit/UIProcess/WebProcessCache.cpp]: That is a good point! I added a static variable holding the result of this call.

I don't believe the behavior is needed in Catalyst.

Thanks for reviewing!

File: Source/WebKit/UIProcess/WebProcessCache.cpp [hunk 13]
```diff
@@ -59,8 +63,38 @@ WebProcessCache::WebProcessCache(WebProcessPool& processPool)
...
+        WEBPROCESSCACHE_RELEASE_LOG("canCacheProcess: Not caching process because there is not enough free memory, free RAM: %lu", process.processID(), WTF::ramUnused());
+        return false;
+    }
+
+    if (process.hasMediaActivity()) {
+        WEBPROCESSCACHE_RELEASE_LOG("canCacheProcess: Not caching process because there is media activity.", process.processID());
+        return false;
+    }
+
+    if (checkSuspendedState == CheckSuspendedState::Yes && process.throttler().currentState() != ProcessThrottleState::Suspended) {
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebProcessCache.cpp]: I don't understand this check. The process might be in the middle of preparing for suspension and we would discard it as a result. This seems unfortunate.

File: Source/WebKit/UIProcess/WebProcessCache.cpp [hunk 14]
```diff
@@ -91,7 +125,7 @@ bool WebProcessCache::addProcessIfPossible(Ref<WebProcessProxy>&& process)
     ASSERT(!process->suspendedPageCount());
     ASSERT(!process->isRunningServiceWorkers());
 
-    if (!canCacheProcess(process))
+    if (!canCacheProcess(process, CheckSuspendedState::Yes))
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebProcessCache.cpp]: I don't like this. This adds complexity and it is unclear why we need it.

File: Source/WebKit/UIProcess/WebProcessCache.cpp [hunk 15]
```diff
@@ -104,7 +138,6 @@ bool WebProcessCache::addProcessIfPossible(Ref<WebProcessProxy>&& process)
         auto cachedProcess = m_pendingAddRequests.take(requestIdentifier);
         if (!cachedProcess)
             return;
-
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebProcessCache.cpp]: Unrelated change.

File: Source/WebKit/UIProcess/WebProcessCache.cpp [hunk 16]
```diff
@@ -174,24 +208,41 @@ RefPtr<WebProcessProxy> WebProcessCache::takeProcess(const WebCore::RegistrableD
...
 {
-    if (!processPool.configuration().processSwapsOnNavigation() || !processPool.configuration().usesWebProcessCache() || LegacyGlobalSettings::singleton().cacheModel() != CacheModel::PrimaryWebBrowser || processPool.configuration().usesSingleWebProcess()) {
+#if PLATFORM(IOS)
+    static bool hasEnabledProcessCache = false;
+    static bool isFullWebBrowser = isFullWebBrowserOrRunningTest();
+    bool usesWebProcessCache = isFullWebBrowser && !hasEnabledProcessCache;
+#elif PLATFORM(MAC)
+    bool usesWebProcessCache = processPool.configuration().usesWebProcessCache();
+#else
+    bool usesWebProcessCache = false;
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebProcessCache.cpp]: Why would we ignore `processPool.configuration().usesWebProcessCache()` on iOS?

File: Source/WebKit/UIProcess/WebProcessProxy.h [hunk 17]
```diff
@@ -495,6 +498,8 @@ class WebProcessProxy : public AuxiliaryProcessProxy, private ProcessThrottlerCl
 
     Logger& logger();
 
+    bool hasMediaActivity() const { return m_audibleMediaActivity.has_value() || m_mediaStreamingActivity.has_value(); }
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebProcessProxy.h]: This feels like a hack? As I remember it, we drop those activities on a timer (in case new media starts soon after).
If a process goes into the process cache, we should just drop those media activities, not throw away the whole process.

File: Source/WebKit/WebProcess/WebProcess.cpp [hunk 18]
```diff
@@ -1594,6 +1595,8 @@ void WebProcess::prepareToSuspend(bool isSuspensionImminent, MonotonicTime estim
     SetForScope suspensionScope(m_isSuspending, true);
     m_processIsSuspended = true;
 
+    parentProcessConnection()->send(Messages::WebProcessProxy::UpdateMemoryUsage(memoryFootprint()), 0);
```
Comment (cdumez) [inline Source/WebKit/WebProcess/WebProcess.cpp]: This is very fragile. The UIProcess has a way to ask the memory footprint of a WebProcess but the value is only accurate if the WebProcess got a PrepareToSuspend IPC. This doesn't feel like the right way to implement this.

Also, there is a conflict in logic:
1. We purposefully don't trim memory in WebProcesses on suspension to get more warm caches when the process gets re-used from the cache, thus making the WebProcess gets larger.
2. But, we refuse to cache a process if it is too large.

This doesn't make much sense. Seems having a cached process without warm cache would still be more valuable than no cached process at all.

File: Source/WebKit/WebProcess/WebProcess.cpp [hunk 19]
```diff
@@ -1609,7 +1612,7 @@ void WebProcess::prepareToSuspend(bool isSuspensionImminent, MonotonicTime estim
         platformMediaSessionManager->processWillSuspend();
 #endif
 
-#if !PLATFORM(MAC)
+#if !PLATFORM(MAC) && !PLATFORM(IOS)
```
Comment (cdumez) [inline Source/WebKit/WebProcess/WebProcess.cpp]: Doesn't this disable process-slimming on iOS before suspension? This doesn't seem like an acceptable / shippable change as it would result in more background WebContent jetsams, which would be a net user regression.

File: Source/WebKit/UIProcess/WebProcessCache.cpp [hunk 20]
```diff
@@ -59,8 +63,38 @@ WebProcessCache::WebProcessCache(WebProcessPool& processPool)
...
 {
+    WEBPROCESSCACHE_RELEASE_LOG("canCacheProcess: memory usage: %lu", process.processID(), process.memoryUsage());
+
+#if PLATFORM(IOS)
+    if (WTF::ramUnused() < 256 * MB) {
+        WEBPROCESSCACHE_RELEASE_LOG("canCacheProcess: Not caching process because there is not enough free memory, free RAM: %lu", process.processID(), WTF::ramUnused());
+        return false;
+    }
+
+    if (process.hasMediaActivity()) {
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebProcessCache.cpp]: A Media activity is just a process assertion. A process wouldn't be able to enter the process cache if it still had media playing in it. Therefore, this check doesn't make much sense.

File: Source/WebKit/UIProcess/WebProcessCache.cpp [hunk 21]
```diff
@@ -59,8 +63,38 @@ WebProcessCache::WebProcessCache(WebProcessPool& processPool)
...
+        return false;
+    }
+
+    if (!process.memoryUsage()) {
+        WEBPROCESSCACHE_RELEASE_LOG("canCacheProcess: Not caching process because the process' memory usage is unknown.", process.processID());
+        return false;
+    }
+
+    constexpr unsigned memoryLimit = 100 * MB;
+    if (process.memoryUsage() > memoryLimit) {
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebProcessCache.cpp]: This check doesn't make a lot of sense, see comments below.

**Review by cdumez — CHANGES_REQUESTED** [review]

File: Source/WebKit/UIProcess/WebProcessCache.cpp (hunk 21 above)
Comment (pvollan) [reply Source/WebKit/UIProcess/WebProcessCache.cpp]: > This check doesn't make a lot of sense, see comments below.

Thanks for reviewing, Chris! I will address these issues in an upcoming patch :)

File: Source/WebKit/UIProcess/WebProcessCache.cpp [hunk 22]
```diff
@@ -131,9 +135,15 @@ bool WebProcessCache::addProcess(std::unique_ptr<CachedProcess>&& cachedProcess)
         WEBPROCESSCACHE_RELEASE_LOG("addProcess: Evicting process from WebProcess cache because a new process was added for the same domain", previousProcess->process().processID());
 
     while (m_processesPerRegistrableDomain.size() >= capacity()) {
-        auto it = m_processesPerRegistrableDomain.random();
-        WEBPROCESSCACHE_RELEASE_LOG("addProcess: Evicting process from WebProcess cache because capacity was reached", it->value->process().processID());
-        m_processesPerRegistrableDomain.remove(it);
+        // Remove the process using the most memory
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebProcessCache.cpp]: I don't think I agree with this change. Random eviction is our preferred cache eviction policy. Why evict a cached process simply because it is bigger? Cached processes go away on memory pressure anyway. Also, a large process means a process which populated caches with means the page load will likely be faster if we use it.

File: Source/WebKit/UIProcess/WebProcessProxy.cpp [hunk 23]
```diff
@@ -1377,6 +1378,25 @@ void WebProcessProxy::renderTreeAsText(WebCore::ProcessIdentifier destinationPro
 
 bool WebProcessProxy::canBeAddedToWebProcessCache() const
 {
+    // Don't cache WebContent processes that are running. Before being considered for caching,
+    // WebContent processes will be suspended and inserted into the Back/Forward cache.
+    // If the process has not suspended by now, there is likely an assertion leak, that keeps
+    // it running. It is better to terminate the process, than keep it running by adding it to
+    // the cache.
+    if (throttler().currentState() != ProcessThrottleState::Suspended) {
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebProcessProxy.cpp]: As mentioned in my earlier review, I disagree with this. Why aren't we caching processes that are not suspended? This seems racy. Processes enter the process cache when they are no longer used but getting suspended is something that happens asynchronously and may take an arbitrary amount of time.

File: Source/WebKit/UIProcess/WebProcessProxy.cpp [hunk 24]
```diff
@@ -2526,6 +2549,12 @@ void WebProcessProxy::addAllowedFirstPartyForCookies(const WebCore::RegistrableD
     send(Messages::WebProcess::AddAllowedFirstPartyForCookies(firstPartyDomain), 0);
 }
 
+void WebProcessProxy::releaseMemoryAfterSuspensionTimerFired()
+{
+    auto activity = throttler().backgroundActivity("Release memory after suspension"_s);
+    send(Messages::WebProcess::ReleaseMemory({ }), 0);
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebProcessProxy.cpp]: The activity goes out of scope right away and thus the process likely doesn't stay awake long enough to process the IPC.

Comment (cdumez) [reply Source/WebKit/UIProcess/WebProcessProxy.cpp]: Also, we don't want to wake up suspended processes in general. This will page them in memory again.

File: Source/WebKit/WebProcess/WebProcess.cpp [hunk 25]
```diff
@@ -1609,14 +1610,6 @@ void WebProcess::prepareToSuspend(bool isSuspensionImminent, MonotonicTime estim
         platformMediaSessionManager->processWillSuspend();
 #endif
 
-#if !PLATFORM(MAC)
```
Comment (cdumez) [inline Source/WebKit/WebProcess/WebProcess.cpp:1612]: Processes on iOS now no longer slim down on suspension and we thus have a higher background jetsam risk, as mentioned in my earlier review.

File: Source/WebKit/NetworkProcess/NetworkProcess.cpp [hunk 26]
```diff
@@ -2249,7 +2250,7 @@ void NetworkProcess::prepareToSuspend(bool isSuspensionImminent, MonotonicTime e
     RefPtr<CallbackAggregator> callbackAggregator = CallbackAggregator::create([this, completionHandler = WTFMove(completionHandler)]() mutable {
         RELEASE_LOG(ProcessSuspension, "%p - NetworkProcess::prepareToSuspend() Process is ready to suspend", this);
         UNUSED_VARIABLE(this);
-        completionHandler();
+        completionHandler(memoryFootprint());
```
Comment (cdumez) [inline Source/WebKit/NetworkProcess/NetworkProcess.cpp]: This looks very weird. We only make use of this information for the WebProcesses and yet require all child processes to compute it and send it back, at a time where we want the process to suspend ASAP.

**Review by cdumez — CHANGES_REQUESTED** [review]

File: Source/WebKit/NetworkProcess/NetworkProcess.cpp (hunk 26 above)
Comment (pvollan) [reply Source/WebKit/NetworkProcess/NetworkProcess.cpp]: Thanks for reviewing, Chris! I will address these issues in a follow-up patch :)

File: Source/WebKit/UIProcess/WebProcessCache.cpp (hunk 22 above)
Comment (pvollan) [reply Source/WebKit/UIProcess/WebProcessCache.cpp]: That is a good point! Although, I also think evicting by memory size has some good features. For example, it will minimize the total memory footprint for a given cache size, and thereby minimizing the chance for jetsams. Another benefit is that the behavior will be more consistent, which will reduce the variance in benchmark runs.

Comment (cdumez) [reply Source/WebKit/UIProcess/WebProcessCache.cpp]: > For example, it will minimize the total memory footprint for a given cache size, and thereby minimizing the chance for jetsams. 
Cached processes get cleared on memory pressure already.

File: Source/WebKit/UIProcess/WebProcessProxy.cpp (hunk 23 above)
Comment (pvollan) [reply Source/WebKit/UIProcess/WebProcessProxy.cpp]: Ah, yes, good catch, this is racy. I will attempt to address this in the next patch.

File: Source/WebKit/UIProcess/WebProcessCache.cpp (hunk 22 above)
Comment (cdumez) [reply Source/WebKit/UIProcess/WebProcessCache.cpp]: Also, this patch adds a lot of complexity just to determine the process size and requires processes to be suspended to enter the cache. This is not a good trade-off.

File: Source/WebKit/UIProcess/WebProcessProxy.cpp (hunk 24 above)
Comment (pvollan) [reply Source/WebKit/UIProcess/WebProcessProxy.cpp]: Ah, good catch! Will fix in upcoming patch. In this case, perhaps it would be ok to wake up the process, since we will only run a limited amount of code to free memory, which should not touch many pages?

Comment (cdumez) [reply Source/WebKit/UIProcess/WebProcessProxy.cpp]: I don't think we can do that. It WILL page a lot of stuff in and increase jetsam priority.
 Feel free to talk to @bnham about it too. I remember he was opposed to doing this in the past for the reasons I mentioned.

File: Source/WebKit/WebProcess/WebProcess.cpp (hunk 25 above)
Comment (pvollan) [reply Source/WebKit/WebProcess/WebProcess.cpp:1612]: Yes, that is a good point. Page load performance benefits from having this cached when the process is taken from the process cache. I wonder if we could delay releasing this memory until the very last assertion is dropped, since the process will be able to respond to memory pressure warnings before that?

File: Source/WebKit/WebProcess/WebProcess.cpp [hunk 27]
```diff
@@ -447,7 +447,7 @@ void WebProcess::initializeWebProcess(WebProcessCreationParameters&& parameters)
             if (m_pagesInWindows.isEmpty() && critical == Critical::No)
                 critical = Critical::Yes;
 
-#if PLATFORM(MAC)
+#if PLATFORM(MAC) || PLATFORM(IOS_FAMILY)
```
Comment (brentfulgham) [inline Source/WebKit/WebProcess/WebProcess.cpp]: If this is Mac + all iOS things, maybe it should just be `#if PLATFORM(COCOA)`?

**Review by brentfulgham — APPROVED** [review]
r=me

File: Source/WebKit/WebProcess/WebProcess.cpp (hunk 27 above)
Comment (pvollan) [reply Source/WebKit/WebProcess/WebProcess.cpp]: Ah, good catch, will fix!

Thanks for reviewing!

File: Source/WTF/wtf/RAMSize.cpp (hunk 12 above)
Comment (pvollan) [reply Source/WTF/wtf/RAMSize.cpp]: Ah, good point! This function has been removed in the latest patch.
